atmos_model.F90

```diff
@@ -929,49 +935,113 @@ subroutine get_atmos_model_ungridded_dim(nlev, nsoillev, ntracers,     &
+!            "longname",     "PM2.5"
+!            "units",        "ug/m3"
+!            "tracer_usage", "chemistry", "type=diagnostic"
+!            "profile_type", "fixed", "surface_value=5.e-6" /
+!
+!  For atmospheric chemistry, the order of both prognostic and diagnostic
+!  tracers is validated against the model's internal assumptions.
+!
+! </DESCRIPTION>
+subroutine get_atmos_tracer_types(tracer_types)
+
+  use field_manager_mod,  only: parse
+  use tracer_manager_mod, only: query_method
+
+  integer, intent(out) :: tracer_types(:)
+
+  !--- local variables
+  logical :: found
+  integer :: n, num_tracers, num_types
+  integer :: id_max, id_min, id_num, ip_max, ip_min, ip_num
+  character(len=32)  :: tracer_usage
+  character(len=128) :: control, tracer_type
+
+  !--- begin
+
+  !--- validate array size
+  call get_number_tracers(MODEL_ATMOS, num_tracers=num_tracers)
+
+  if (size(tracer_types) < num_tracers) &
+    call mpp_error(FATAL, 'insufficient size of tracer type array')
+
+  !--- initialize tracer indices
+  id_min = num_tracers + 1
+  id_max = -id_min
+  ip_min = id_min
+  ip_max = id_max
+  id_num = 0
+  ip_num = 0
+
+  do n = 1, num_tracers
+    tracer_types(n) = 0
+    found = query_method('tracer_usage',MODEL_ATMOS,n,tracer_usage,control)
+    if (found) then
+      if (trim(tracer_usage) == 'chemistry') then
+        !--- set default to prognostic
+        tracer_type = 'prognostic'
+        num_types = parse(control, 'type', tracer_type)
+        select case (trim(tracer_type))
+          case ('diagnostic')
+            tracer_types(n) = 2
+            id_num = id_num + 1
+            id_max = n
+            if (id_num == 1) id_min = n
+          case ('prognostic')
+            tracer_types(n) = 1
+            ip_num = ip_num + 1
+            ip_max = n
+            if (ip_num == 1) ip_min = n
+        end select
+      end if
     end if
-  end if
+  end do
 
-  !--- number of bins for chemistry diagnostic output
-  if (present(num_diag_sfc_emis_flux)) then
-    num_diag_sfc_emis_flux = 0
-    if (associated(GFS_data(1)%IntDiag%duem)) &
-      num_diag_sfc_emis_flux = size(GFS_data(1)%IntDiag%duem, dim=2)
-    if (associated(GFS_data(1)%IntDiag%ssem)) &
-      num_diag_sfc_emis_flux = &
-        num_diag_sfc_emis_flux + size(GFS_data(1)%IntDiag%ssem, dim=2)
+  if (ip_num > 0) then
+    !--- check if prognostic tracers are contiguous
+    if (ip_num > ip_max - ip_min + 1) &
+      call mpp_error(FATAL, 'prognostic chemistry tracers must be contiguous')
   end if
 
-  !--- number of tracers used in emission diagnostic output
-  if (present(num_diag_burn_emis_flux)) then
-    num_diag_burn_emis_flux = 0
-    if (associated(GFS_data(1)%IntDiag%abem)) &
-      num_diag_burn_emis_flux = size(GFS_data(1)%IntDiag%abem, dim=2)
+  if (id_num > 0) then
+    !--- check if diagnostic tracers are contiguous
+    if (id_num > id_max - id_min + 1) &
+      call mpp_error(FATAL, 'diagnostic chemistry tracers must be consecutive')
```

Comment:
Change to `contiguous` for consistency with line 1031?

Reply:
Done in commit c48922a